ext/drreg/drreg.c

```diff
@@ -1719,37 +2714,82 @@ drreg_event_restore_state(void *drcontext, bool restore_memory,
         return true; /* fault not in cache */
     for (reg = DR_REG_START_GPR; reg <= DR_REG_STOP_GPR; reg++)
         spilled_to[GPR_IDX(reg)] = MAX_SPILLS;
+#ifdef SIMD_SUPPORTED
+    for (reg = DR_REG_APPLICABLE_START_SIMD; reg <= DR_REG_APPLICABLE_STOP_SIMD; reg++)
+        spilled_simd_to[SIMD_IDX(reg)] = MAX_SIMD_SPILLS;
+    for (slot = 0; slot < ops.num_spill_simd_slots; ++slot)
+        simd_slot_use[slot] = DR_REG_NULL;
+#endif
+
     LOG(drcontext, DR_LOG_ALL, 3,
         "%s: processing fault @" PFX ": decoding from " PFX "\n", __FUNCTION__,
         info->raw_mcontext->pc, pc);
     instr_init(drcontext, &inst);
+    instr_init(drcontext, &next_inst);
     while (pc < info->raw_mcontext->pc) {
         instr_reset(drcontext, &inst);
+        instr_reset(drcontext, &next_inst);
         prev_pc = pc;
         pc = decode(drcontext, pc, &inst);
-
-        if (is_our_spill_or_restore(drcontext, &inst, &spill, &reg, &slot, &offs)) {
+        if (pc == NULL) {
+            LOG(drcontext, DR_LOG_ALL, 3, "%s @" PFX " \n", __FUNCTION__, prev_pc,
+                "PC decoding returned NULL during state restoration");
+            instr_free(drcontext, &inst);
+            instr_free(drcontext, &next_inst);
+            return true;
+        }
+        decode(drcontext, pc, &next_inst);
+        if (is_our_spill_or_restore(drcontext, &inst, &next_inst, &is_spill, &reg, &slot,
+                                    &offs, &is_indirect_spill)) {
             LOG(drcontext, DR_LOG_ALL, 3,
                 "%s @" PFX " found %s to %s offs=0x%x => slot %d\n", __FUNCTION__,
-                prev_pc, spill ? "spill" : "restore", get_register_name(reg), offs, slot);
-            if (spill) {
+                prev_pc, is_spill ? "is_spill" : "restore", get_register_name(reg), offs,
+                slot);
+            if (is_spill) {
                 if (slot == AFLAGS_SLOT) {
                     spilled_to_aflags = slot;
+#ifdef SIMD_SUPPORTED
+                } else if (is_indirect_spill) {
+                    if (reg_is_vector_simd(reg)) {
```

Comment:
Should this be an assert instead?

Reply:
This was aimed so that the handling of additional register classes can be implemented by extending the branch with an if else. Remember the code was designed with flexibility in mind, hence initially avoiding hard asserts on XMM spillage. 

Of course, I'll fix this as suggested.

Reply:
Done.